File: LeetCode/Problems/287.py
# FLOYD'S TURTLE AND HARE ALGORITHM

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow = 0
        fast = 0
        while slow != fast or slow == fast == 0:
            slow = nums[slow]
            fast = nums[nums[fast]]

        slow2 = 0
        while slow2 != slow:
            slow2 = nums[slow2]
            slow = nums[slow]

        return slow

File: leetcode-new/287.cpp
class Solution
{
public:
    int findDuplicate(vector<int> &nums)
    {
        int n = nums.size();
        int r = n - 1;
        int l = 0;
        while (l < r)
        {
            int m = (l + r) >> 1;
            int count = 0;
            for (int i = 0; i < n; i++)
                if (nums[i] <= m)
                    ++count;
            if (count > m)
                r = m;
            else
                l = m + 1;
        }
        return l;
    }
};
